### ISSUE TITLE
Query on Gradient accumulation

### ISSUE THREAD

**Issue by Vattikondadheeraj** [pytorch/torchtune#2134]
Hey, I have one doubt on gradient accumulation parameter. When I increase the parameter from 4 to 8, I am getting OOM error which doesn't make much sense to me. I just wanna ask why am I getting this error? Are you storing the gradients individually or summing them as we get new ones? Or Am i missing something else?

**Comment by mnmnk43434** [comment]
Hello, we tried to solve the issue.

This is what we did:

Update the train method in FullFinetuneRecipeDistributed to properly handle gradient accumulation without increasing memory usage. This involves modifying how gradients are accumulated and when the optimizer step is performed.

You can review changes in this commit: https://github.com/mnmnk43434/pytorch-torchtune-2134/commit/a071b3e5f45a66877db29d193f5b4d783e88fcc3.


> [!CAUTION]
> Disclaimer: The concept of solution was created by AI and you should never copy paste this code before you check the correctness of generated code. Solution might not be complete, you should use this code as an inspiration only.

---

If you no longer want Latta AI to attempt solving issues on your repository, you can block this account.

**Comment by ebsmothers** [comment]
Hi @Vattikondadheeraj thanks for creating the issue. This is a bit surprising to me. We call `.backward()` [after every iteration](https://github.com/pytorch/torchtune/blob/06a837953a89cdb805c7538ff5e0cc86c7ab44d9/recipes/full_finetune_distributed.py#L841) so there should not be a memory increase when you increase the gradient accumulation parameter. Can you share repro commands so that we can debug?
